### ISSUE TITLE
Display oddity

### ISSUE THREAD

**Issue by ArjunNair** [Dotneteer/spectnetide#107]
![specnetide](https://user-images.githubusercontent.com/1013419/52340344-897fd400-2a35-11e9-8a63-87d899e8c932.png)

I'm not sure if it's just me, but there appears to be something funny going with the pixels over the copyright symbol. It seems to be related to issue 72, but in my case the emulator isn't paused or anything.

**Comment by Dotneteer** [comment]
@ArjunNair, can you reproduce it? If yes, please zip the entire solution folder and upload it to this issue so that I can analyze it!

**Comment by ArjunNair** [comment]
[SpectrumCodeDiscovery1.zip](https://github.com/Dotneteer/spectnetide/files/2837341/SpectrumCodeDiscovery1.zip)

Yes, I get the same issue every time. I've zipped and attached it here but it's actually the example Code Discovery project that comes with the VISX!

**Comment by Dotneteer** [comment]
Thanks, I will check it soon!

**Comment by Dotneteer** [comment]
@ArjunNair, I wanted to check if the files within the solution are not corrupted, and this why I asked you to upload them.

The solution works perfectly on my machine, it produces the expected screen on my both monitor, so I cannot reproduce the oddity. I guess, there might be issues with something within the XAML bitmap rendering on your machine -- but it is just guessing. 

I saw that the image you've sent to me has a 3x zoom. Not only the copyright symbol is odd, but also digit 1 in 1982 is displayed in an odd form.

Can you try to load and run some program that manipulates/changes the screen and check if you see any other oddity? If you have a second monitor, please try to move the ZX Spectrum emulator screen there, resize it and check if you find this or other oddities. Your experience may give me some clues where to look for the issue.

I do not believe there is any bug in SpectNetIDE causing this phenomenon, but believing does not mean I'm 100% right...

**Comment by ArjunNair** [comment]
Ok, I'll check and get back to you in a bit. Note that I posted with 3x zoom just to show the issue at a meaningful resolution. The issue happens irrespective of the window size, however. I noticed that if I add some bit of code to write pixels to the screen, I don't get the same issue anymore (something to do with HALT and the ULA rendering maybe?). Another odd thing is, that when the emulator starts I can see the image being displayed correctly for 1 frame (maybe more. It's hard to tell.) but it quickly changes to this after a few frames and stays this way. Anyway, let me get back to you with some more info.

**Comment by Dotneteer** [comment]
@ArjunNair, I managed to understand what's going on :-). You actually ran the code in the `Code.z80asm` file :-). This code results the odd screen on my machine, too, so I could reproduce it.
I know why it displays the odd screen, and in a next comment I will explain it to you :-) -- now I need to leave for a customer...

**Comment by Dotneteer** [comment]
@ArjunNair, as I promised, here is the explanation:

When you run the code, SpectNetIDE carries out these steps:
1. Starts a new ZX Spectrum and pauses it when it reaches the __$12A9__ address -- it is the main statement execution loop. (To accelerate the startup, the IDE loads this paused state from the cache, provided the state already exists)
2. Injects the code into the memory.
3. Sets the PC register to the start address of your code
4. Resumes the virtual machine that starts executing your code.

By default, the `Code.z80asm` file contains this code:

```
; Put your Z80 assembly code into code files
Start:
    .org #8000
    halt
```
As you may recognize, this is an incomplete code. The `HALT` statement stops the code only by the time the next interrupt (start of the next frame screen) arrives and then goes on with whatever code is behind the `HALT` instruction.

__Note, this is why your first screen frame does not have the oddity.__

When code overflows the `HALT` instruction at __$8000__, there are a lot of zeros in the memory from __$8001__ to __$FF38__, when the extent of the the stack so far begins. The CPU starts to execute the junk bytes -- the contents of the stack -- believe it or not, this junk code puts the odd bytes into the screen memory :-)
The junk code somehow gets back to the main execution cycle where it waits for a keypress, and then it causes the virtual machine to restart.

Add this code line after the `HALT` instruction:

```
    jp #12a9
```
When you run the code, there's no oddity :-).

Thanks for reporting this issue! I will change the default code to something more useful instead of a single `HALT` instruction, as it may be deceptive for anyone who's novice in Z80 coding.

I will close this issue when I change the default content of `Code.z80asm`.

**Comment by Dotneteer** [comment]
Fixed, will be released soon.

**Comment by ArjunNair** [comment]
Sounds good! Thanks!

**Comment by Dotneteer** [comment]
Released in v1.16.0.

**Comment by Dotneteer** [comment]
@ArjunNair, if you like this project, please star it on Github :-)
